File: python/dssatutils/weather_rainfall_merge.py
"""Helpers for replacing coarse weather-source rainfall with gridded rain."""

from __future__ import annotations

from typing import Mapping

import pandas as pd


def merge_rainfall_into_weather(
    weather: pd.DataFrame,
    rainfall: Mapping[str, float] | pd.Series | None,
    date_col: str = "DATE",
    rain_col: str = "RAIN",
) -> int:
    """Replace *weather[rain_col]* where daily rainfall has matching DATE keys.

    Parameters
    ----------
    weather:
        Daily weather frame with DSSAT-style ``YYYYDOY`` date codes.
    rainfall:
        Mapping or Series keyed by ``YYYYDOY`` with rainfall in mm/day.
    date_col, rain_col:
        Column names in *weather*.

    Returns
    -------
    int
        Number of daily rows replaced. The input DataFrame is modified in place.
    """
    if rainfall is None:
        return 0
    series = pd.Series(rainfall, dtype="float64")
    if series.empty:
        return 0
    series.index = series.index.astype(str)
    mapped = weather[date_col].astype(str).map(series)
    use = mapped.notna()
    if use.any():
        weather[rain_col] = pd.to_numeric(weather[rain_col], errors="coerce").astype("float64")
        weather.loc[use, rain_col] = mapped[use].values
    return int(use.sum())

File: R/weather_rainfall_merge.R
# Shared helper: replace a weather frame's RAIN column from daily rainfall.

merge_rainfall_into_weather <- function(weather_data, rainfall,
                                        date_col = "DATE", rain_col = "RAIN") {
  if (is.null(rainfall) || length(rainfall) == 0) {
    return(list(weather_data = weather_data, n_replaced = 0L))
  }
  if (is.null(names(rainfall))) {
    stop("rainfall must be a named vector keyed by DSSAT DATE codes (YYYYDOY).",
         call. = FALSE)
  }
  idx <- match(as.character(weather_data[[date_col]]), names(rainfall))
  hit <- !is.na(idx)
  if (any(hit)) {
    weather_data[[rain_col]] <- as.numeric(weather_data[[rain_col]])
    weather_data[[rain_col]][hit] <- as.numeric(rainfall[idx[hit]])
  }
  list(weather_data = weather_data, n_replaced = sum(hit))
}
